pkg/core/native/notary.go

```diff
@@ -0,0 +1,345 @@
+		manifest.NewParameter("addr", smartcontract.Hash160Type),
+		manifest.NewParameter("till", smartcontract.IntegerType))
+	md = newMethodAndPrice(n.lockDepositUntil, 100_0000, smartcontract.AllowStates)
+	n.AddMethod(md, desc, true)
+
+	desc = newDescriptor("withdraw", smartcontract.BoolType,
+		manifest.NewParameter("from", smartcontract.Hash160Type),
+		manifest.NewParameter("to", smartcontract.Hash160Type))
+	md = newMethodAndPrice(n.withdraw, 100_0000, smartcontract.AllowStates)
+	n.AddMethod(md, desc, true)
+
+	desc = newDescriptor("balanceOf", smartcontract.IntegerType,
+		manifest.NewParameter("addr", smartcontract.Hash160Type))
+	md = newMethodAndPrice(n.balanceOf, 100_0000, smartcontract.AllowStates)
+	n.AddMethod(md, desc, true)
+
+	desc = newDescriptor("expirationOf", smartcontract.IntegerType,
+		manifest.NewParameter("addr", smartcontract.Hash160Type))
+	md = newMethodAndPrice(n.expirationOf, 100_0000, smartcontract.AllowStates)
+	n.AddMethod(md, desc, true)
+
+	desc = newDescriptor("verify", smartcontract.BoolType)
+	md = newMethodAndPrice(n.verify, 100_0000, smartcontract.NoneFlag)
+	n.AddMethod(md, desc, false)
+
+	desc = newDescriptor("name", smartcontract.StringType)
+	md = newMethodAndPrice(nameMethod(notaryName), 0, smartcontract.NoneFlag)
+	n.AddMethod(md, desc, true)
+
+	desc = newDescriptor("onPersist", smartcontract.VoidType)
+	md = newMethodAndPrice(getOnPersistWrapper(n.OnPersist), 0, smartcontract.AllowModifyStates)
+	n.AddMethod(md, desc, false)
+
+	desc = newDescriptor("postPersist", smartcontract.VoidType)
+	md = newMethodAndPrice(getOnPersistWrapper(postPersistBase), 0, smartcontract.AllowModifyStates)
+	n.AddMethod(md, desc, false)
+
+	return n
+}
+
+// Metadata implements Contract interface.
+func (n *Notary) Metadata() *interop.ContractMD {
+	return &n.ContractMD
+}
+
+// Initialize initializes Notary native contract and implements Contract interface.
+func (n *Notary) Initialize(ic *interop.Context) error {
+	return nil
+}
+
+// OnPersist implements Contract interface.
+func (n *Notary) OnPersist(ic *interop.Context) error {
+	var nFees int64
+	for _, tx := range ic.Block.Transactions {
+		if tx.HasAttribute(transaction.NotaryAssistedT) {
+			nkeys := tx.GetAttributes(transaction.NotaryAssistedT)[0].Value.(*transaction.NotaryAssisted).NKeys
+			nFees += int64(nkeys) + 1
+		}
+	}
+	if nFees == 0 {
+		return nil
+	}
+	notaries, err := n.GetNotaryNodes(ic.DAO)
+	if err != nil {
+		return fmt.Errorf("failed to get notary nodes: %w", err)
+	}
+	singleReword := calculateNotaryReword(nFees, len(notaries))
+	for _, notary := range notaries {
+		n.GAS.mint(ic, notary.GetScriptHash(), singleReword)
+	}
+	return nil
+}
+
+// onPayment records deposited amount as belonging to "from" address with a lock
+// till the specified chain's height.
+func (n *Notary) onPayment(ic *interop.Context, args []stackitem.Item) stackitem.Item {
+	from := toUint160(args[0])
+	amount := toBigInt(args[1])
+
```

Comment:
You need to check the calling script hash here first of all, it should be GAS (otherwise abort the execution).